agent/xds/clusters.go

```diff
@@ -449,3 +455,27 @@ func (s *Server) makeMeshGatewayCluster(clusterName string, cfgSnap *proxycfg.Co
 		OutlierDetection: &envoycluster.OutlierDetection{},
 	}, nil
 }
+
+func makeThresholdsIfNeeded(limits UpstreamLimits) []*envoycluster.CircuitBreakers_Thresholds {
+	var empty UpstreamLimits
+	// Make sure to not create any thresholds when passed the zero-value in order
+	// to rely on Envoy defaults
+	if limits == empty {
+		return nil
+	}
+
+	threshold := &envoycluster.CircuitBreakers_Thresholds{}
+	// Likewise, make sure to not set any threshold values on the zero-value in
+	// order to rely on Envoy defaults
+	if limits.MaxConnections != 0 {
```

Comment:
I wonder if it's ever valid in Envoy to want to explicitly set these to `0` to disable them or similar? If so we'd need something a little more sophisticated to be able to disambiguate "not set" and "explicit zero"...

At any rate it's likely an edge case but perhaps `-1` could be purposed for this?

Reply:
An interesting thought experiment around why someone might want to set that, but I agree we should not make it impossible for someone to do that.

Instead of `-1` as a sentinel value, I am slightly inclined to use `*int` types and represent `nil` as the zero value. Not sure if we have any standards around this elsewhere.

Reply:
We use pointers in the config package so we can internally distinguish between not set and zero, but we don't have that anywhere in public API currently. I'm not quite sure how it would work in mapstructure although it probably would I guess and it would therefore be "invisible" to users which is nice so +1.

Where it does end up being visible, we have at least some precedent for using `-1` for disable in our `ports{}` config. In that case it wasn't because of the default zero thing since config already uses pointers, it was because port `0` is actually a valid thing too (meaning let the OS pick a free one for me).

Honestly, I'm on the fence about whether it's even worth supporting yet especially given these specific parameters - setting a really high number like 1000000 is also a fine work around if you really don't want to choose a sensible limit anyway. But yeah pointer would be cleaner in this case.

Reply:
Yes, I think using pointer values is pretty clean and sets a nice precedent for any additional properties that are added.

Also, to be clear, the Envoy default values if something is not set is actually `1024` for all three configuration values, not unlimited. So setting something really high like 1000000 is actually an explicit choice away from the default.

Reply:
Does Envoy discriminate between explicit `0` and unset? If no this is a moot conversation anyway!

But glad we came up with a clean solution in general 😄?

Reply:
Good question. Did a quick test and envoy does let you set `max_connections: 0` and then not allow any connections to be made to the upstream. 👍